src/test/components/Root-history.test.js

```diff
@@ -0,0 +1,185 @@
+        getProfileFromTextSamples('A  B  C  D  E').profile
+      );
+      mockFetchProfileAtUrl(getProfileUrlForHash('FAKEHASH'), profile);
+    } else {
+      throw new Error('TODO');
+    }
+
+    const ctx = mockCanvasContext();
+    jest
+      .spyOn(HTMLCanvasElement.prototype, 'getContext')
+      .mockImplementation(() => ctx);
+
+    const store = blankStore();
+    const renderResult = render(<Root store={store} />);
+
+    _cleanup = () => {
+      // Cleanup the mocks.
+      resetWindowLocation();
+      resetWindowHistory();
+      delete window.fetch;
+      _cleanup = null;
+    };
+
+    const { findByRole } = renderResult;
+
+    async function waitForTab(details: *): Promise<HTMLElement> {
+      return findByRole('tab', details);
+    }
+
+    return {
+      ...renderResult,
+      ...store,
+      waitForTab,
+    };
+  }
+
+  it('can view a file from the profile store, use history with it', async function() {
+    const { getByText, queryByText, waitForTab } = setup({
+      profileHash: 'FAKEHASH',
+    });
+
+    expect(window.history.length).toBe(1);
+
+    expect(getByText('Downloading and processing the profile...')).toBeTruthy();
+    expect(queryByText('Call Tree')).toBeFalsy();
+
+    await Promise.all((window: any).fetch.mock.results.map(n => n.value));
+
+    // Wait until the call tree is visible.
+    await waitForTab({ name: 'Call Tree', selected: true });
+    await waitForTab({ name: 'Marker Chart', selected: false });
+
+    // History on load starts at 1.
+    expect(window.history.length).toBe(1);
+
+    // Going back doesn't do anything.
+    window.history.back();
+    expect(window.history.length).toBe(1);
+
+    // Trigger a history event by clicking a tab.
+    const markerChart = await waitForTab({
+      name: 'Marker Chart',
+      selected: false,
+    });
+    markerChart.click();
+
+    await waitForTab({ name: 'Call Tree', selected: false });
+    await waitForTab({ name: 'Marker Chart', selected: true });
+
+    expect(window.history.length).toBe(2);
+
+    // Now go back to the call tree.
+    window.history.back();
+
+    await waitForTab({ name: 'Call Tree', selected: true });
+    await waitForTab({ name: 'Marker Chart', selected: false });
+
+    // The history will still have the same length, as we haven't overwritten it.
+    expect(window.history.length).toBe(2);
+  });
+
+  it('resets the history length between tests', async function() {
+    setup({
+      profileHash: 'FAKEHASH',
+    });
+    expect(window.history.length).toBe(1);
+  });
+});
+
+function mockFetchProfileAtUrl(
+  url: string,
+  profile: SerializableProfile
+): void {
+  const responses = [];
+  const fetch = jest.fn().mockImplementation((fetchUrl: string) => {
+    if (fetchUrl === url) {
+      const response = coerceMatchingShape<Response>({
+        ok: true,
+        status: 200,
+        headers: coerceMatchingShape<Headers>({
+          get: () => 'application/json',
```

Comment:
I'm somewhat surprised the `coerceMatchingShape` is needed, I didn't use such a thing for example in the test for `shorten-url.test.js` (see the function `mockFetchForBitly`). I'm not opposed though.

Reply:
It's not really needed, but I added it to be a bit more type-safe in my mock, that way what I mocked was at least matching the type definitions we are using.

Reply:
Yeah, I _think_ that this used to work out of the box with `jest.spyOn` but it doesn't seem to be checked now.